lib/matplotlib/sphinxext/plot_directive.py

```diff
@@ -312,9 +321,37 @@ def setup(app):
     app.connect('build-finished', _copy_css_file)
     metadata = {'parallel_read_safe': True, 'parallel_write_safe': True,
                 'version': matplotlib.__version__}
+    app.connect('builder-inited', init_filename_registry)
+    app.add_env_collector(_FilenameCollector)
     return metadata
 
 
+# -----------------------------------------------------------------------------
+# Handle Duplicate Filenames
+# -----------------------------------------------------------------------------
+
+def init_filename_registry(app):
+    env = app.builder.env
+    if not hasattr(env, 'mpl_custom_base_names'):
+        env.mpl_custom_base_names = defaultdict(set)
```

Comment:
The name is a bit generic. Maybe 
```suggestion
        env.mpl_plot_image_basenames = defaultdict(set)
```
Or similar. This naming is *library_directive_option*.